### PR TITLE
feat(integrations): add Temporal Integration (#2572)

### LINKED ISSUE #2572: feat: add Temporal workflow orchestration integration
## Context

Temporal is a fast-growing durable workflow platform used for long-running, fault-tolerant business processes. It is increasingly adopted alongside or instead of Airflow/Prefect for microservice-level orchestration. OpenSRE covers Airflow and Prefect but not Temporal, leaving a gap for teams investigating workflow timeouts, worker failures, or activity retries.

## What to build

- List recent workflow executions with status and failure reason
- Fetch workflow history (events) for a given execution
- List task queues and worker pollers
- Fetch namespace-level metrics (open workflows, activity errors)

## Acceptance Criteria

Follow the [New Integration Checklist](https://github.com/Tracer-Cloud/opensre/blob/main/AGENTS.md#6-new-integration-checklist):

- [ ] Integration config under `app/integrations/temporal.py`
- [ ] Service client under `app/services/temporal/client.py` (Temporal HTTP API or gRPC)
- [ ] Tool(s) under `app/tools/TemporalTool/` with typed input schema
- [ ] Unit/mock tests under `tests/integrations/test_temporal.py`
- [ ] Docs page at `docs/integrations/temporal.mdx` registered in `docs/docs.json`
- [ ] Synthetic or e2e test
- [ ] `make verify-integrations` passes
- [ ] `make lint` and `make typecheck` pass
- [ ] **Video demo (required):** screen recording showing a failed Temporal workflow surfaced in an investigation RCA output — must be attached to the PR before review

### PR DESCRIPTION
**Pull request by kasulani**
Fixes #2572 

#### Describe the changes you have made in this PR

Adds a Temporal workflow-orchestration integration for incident investigation (issue #2572):

  - Service client (app/services/temporal/client.py) — a thin HTTP client over Temporal's frontend REST API (/api/v1/namespaces/{ns}/...), with probe_access() for
  verification.
  - 4 investigation tools — TemporalNamespaceInfoTool, TemporalWorkflowsTool, TemporalWorkflowHistoryTool, TemporalTaskQueueTool. These compose into a natural triage
  drill-down: namespace health → which workflows failed → why (event history) → worker/queue health.
  - Integration wiring — registry spec, setup handler (integrations setup temporal), and verification adapter.
  - Docs (docs/temporal.mdx) + tests — client unit tests, per-tool suites, and a synthetic RCA scenario.

Scope: self-hosted Temporal only (not Temporal Cloud).

This integration uses Temporal's frontend HTTP API `/api/v1/namespaces/{ns}/...`, served on the server's --http-port, e.g. 7243. This HTTP gateway is a self-hosted feature. Temporal Cloud does not expose it — Cloud's data plane is gRPC/mTLS-only (port 7233), and its only HTTP surface is the control-plane Cloud Ops API, which "does not allow interaction with individual Workflows or Activities via HTTP." Cloud support would require a separate gRPC client and is intentionally out of scope for this PR. Verified end-to-end against a local Docker dev server.
 
 

### Demo/Screenshot for feature changes and bug fixes

Temporal HTTP API  - Self hosted (docker)

verification

https://github.com/user-attachments/assets/19ae7767-0561-477b-a0f4-91182d1b16cd

Setup

https://github.com/user-attachments/assets/fa85e41f-f4a5-49bf-aff5-c02a3c0706e1

Investigation

https://github.com/user-attachments/assets/22e6cbc5-174c-4205-b1d6-2fbd244bf057

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

**Problem.** When a Temporal workflow fails, the cause (a failed activity, a downstream error, dead workers, a backed-up task queue) is buried in workflow event history that's tedious to dig through manually. This integration lets the agent pull that evidence directly during an investigation.

**Why HTTP, and why self-hosted only.** Temporal has two interfaces: the native gRPC data-plane API (port 7233, what SDKs use) and a frontend HTTP/JSON gateway `(/api/v1/..., served on --http-port, e.g. 7243)`. I chose HTTP because it's a dependency-light client (httpx, no Temporal SDK/protobuf toolchain) that fits the existing tool pattern. The trade-off, which I verified against Temporal's docs: Temporal Cloud does not expose the frontend HTTP API — Cloud's data plane is gRPC/mTLS-only, and its sole HTTP surface (the Cloud Ops API) is control-plane only and "does not allow interaction with individual Workflows via HTTP." So this client physically cannot reach Cloud. That's a deliberate, documented MVP boundary; Cloud support would need a separate gRPC/mTLS client.

**Key components.** The client exposes one method per tool (get_namespace_info, list_workflow_executions, get_workflow_history, describe_task_queue), each returning a structured dict `({"success": ..., ...} or an error dict)` so tools never raise into the pipeline. Two non-obvious bits live in the client because they're wire-format concerns, not tool logic: `_decode_payload_data` (the HTTP API base64-encodes Payload values — e.g. a status name comes back as IkZhaWxlZCI=) and `_flatten_status_groups` (reshapes nested groupValues into flat `[{status, count}]` so the LLM sees `Failed: 8`, not raw Payload encoding).

**What live verification changed.** Mocks alone weren't enough — running against a real Docker dev server exposed that the HTTP API omits falsy fields `(nextPageToken, archived, pollers)` rather than returning empty values, which caused KeyError crashes. The client now defaults these defensively, and the tests were realigned to the real wire shapes. This is why the PR targets a verified self-hosted server rather than assumed schemas.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

File: app/tools/TemporalNamespaceInfoTool/__init__.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            return {
+                "source": "temporal",
+                "available": True,
+                "name": result["name"],
+                "state": result["state"],
+                "workflow_count": result["workflow_count"],
+                "groups": result["groups"],
+            }
+        finally:
+            client.close()
```
[the remark this answers is not in the document]
Comment (kasulani) [reply app/tools/TemporalNamespaceInfoTool/__init__.py]: This has been addressed

File: app/tools/TemporalTaskQueueTool/__init__.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                }
+            return {
+                "source": "temporal",
+                "available": True,
+                "pollers": result["pollers"],
+                "stats": result["stats"],
+                "total": result["total"],
+            }
+        finally:
+            client.close()
```
[the remark this answers is not in the document]
Comment (kasulani) [reply app/tools/TemporalTaskQueueTool/__init__.py]: This has been addressed

File: app/tools/TemporalWorkflowHistoryTool/__init__.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            return {
+                "source": "temporal",
+                "available": True,
+                "events": result["events"],
+                "total": result["total"],
+                "next_page_token": result["next_page_token"],
+                "archived": result["archived"],
+            }
+        finally:
+            client.close()
```
[the remark this answers is not in the document]
Comment (kasulani) [reply app/tools/TemporalWorkflowHistoryTool/__init__.py]: This has been addressed

File: app/tools/TemporalWorkflowsTool/__init__.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                }
+            return {
+                "source": "temporal",
+                "available": True,
+                "executions": result["executions"],
+                "total": result["total"],
+                "next_page_token": result["next_page_token"],
+            }
+        finally:
+            client.close()
```
[the remark this answers is not in the document]
Comment (kasulani) [reply app/tools/TemporalWorkflowsTool/__init__.py]: This has been addressed

**Comment by kasulani** [comment]
@greptile review

File: app/integrations/config_models.py [hunk 5]
```diff
@@ -925,3 +925,25 @@ class PrefectIntegrationConfig(StrictConfigModel):
...
+    def headers(self) -> dict[str, str]:
+
+        headers = {
+            "Content-Type": "application/json",
+        }
+
+        if self.api_key:
+            headers["Authorization"] = f"Bearer {self.api_key}"
+
+        return headers
```
[the remark this answers is not in the document]
Comment (kasulani) [reply app/integrations/config_models.py:1188]: Thanks — but this matches the established convention in this file rather than diverging from it. StrictConfigModel subclasses here routinely expose a headers property that builds the Authorization header: `DatadogIntegrationConfig`, `VercelIntegrationConfig`, `OpsGenieIntegrationConfig`, `PagerDutyIntegrationConfig`, `IncidentIoIntegrationConfig`, and `AlertmanagerIntegrationConfig` all do this. `TemporalIntegrationConfig.headers` deliberately mirrors `AlertmanagerIntegrationConfig` (Content-Type + optional Bearer).

On serialisation: a Pydantic `@property` being excluded from `model_dump()` is intentional here — it keeps the api-key-bearing header out of any serialised config/state. So I'm leaving this as-is for consistency with the six peer models. Please remember this pattern for future integration configs in this repo.

**Comment by kasulani** [comment]
Hi @muddlebee, this PR is ready for review. quick scoping question on the acceptance criteria before review:

The criteria list the service client as "Temporal HTTP API or gRPC." I implemented the HTTP API client. One consequence worth surfacing, since it isn't obvious from the "or": these aren't equivalent in coverage. 

- The frontend HTTP API is a self-hosted feature. It's what this PR targets, verified end-to-end against a local Docker Temporal server. 

- Temporal Cloud does not expose it — Cloud's data plane is gRPC/mTLS-only (the only HTTP surface Cloud offers is the control-plane Ops API, which per Temporal's docs "does not allow interaction with individual Workflows via HTTP").  

So as it stands, this integration works for self-hosted Temporal but not Temporal Cloud. Covering Cloud would require a separate gRPC/mTLS client.  Is HTTP-only (self-hosted) the intended scope for this issue, or did you expect a gRPC client in this PR so Cloud users are covered too? 

My recommendation is to keep this as a self-hosted MVP and  open a follow-up for gRPC/Cloud. 

Happy to either keep this as a self-hosted MVP and open a follow-up for gRPC/Cloud, or fold gRPC in here if that's the expectation.

**Comment by kasulani** [comment]
> The one defect is an accidental one-line deletion in the SigNoz handler that strips the DB record reference from every SigNoz config loaded from the database going forward. It is unrelated to Temporal but was introduced in the same diff hunk.
> 
> app/integrations/_catalog_impl.py — the SigNoz classification block lost its integration_id assignment; all other files look correct.

I have restored the one-line deletion in the SigNoz handler.

**Comment by kasulani** [comment]
@greptile review

**Comment by kasulani** [comment]
@greptile review

**Review by Devesh36 — APPROVED** [review]

**Comment by Devesh36** [comment]
cc @muddlebee

**Comment by kasulani** [comment]
@greptile review

**Comment by muddlebee** [comment]
did you add for? https://www.opensre.com/docs/integrations-overview#local-cli-setup

```
opensre integrations setup
```

and entries missing in .env.example for the relevant env stuff

and merge conflicts

**Comment by kasulani** [comment]
> did you add for? https://www.opensre.com/docs/integrations-overview#local-cli-setup
> 
> ```
> opensre integrations setup
> ```
> 
> and entries missing in .env.example for the relevant env stuff
> 
> and merge conflicts

Thanks @muddlebee for the feedback. I have fixed all the issues you have raised in this comment. Please review again.

**Comment by kasulani** [comment]
@greptile review

**Comment by muddlebee** [comment]
@kasulani LGTM nice work

**Comment by kasulani** [comment]
> @kasulani LGTM nice work

@muddlebee should I create a follow-up issue to extend the temporal integration and add cloud integration as I suggested in [this comment](https://github.com/Tracer-Cloud/opensre/pull/2805#issuecomment-4764393560)?

**Comment by muddlebee** [comment]
@kasulani yes you can I have to explore it more. Don't create PR for now.
